week_4/class_rational_part_1/class_rational_part_1.cpp

```diff
@@ -0,0 +1,120 @@
+#include <iostream>
+
+class Rational {
+ public:
+  Rational() {
+    p = 0;
+    q = 1;
+    divider = 1;
+  }
+
+  Rational(int numerator, int denominator) {
+    if (numerator == 0) {
+      p = numerator;
+      q = 1;
+    }
+
+    else if (numerator < 0 && denominator < 0) {
+      p = abs(numerator);
+      q = abs(denominator);
+    } else if (numerator < 0 || denominator < 0) {
+      p = abs(numerator) * (-1);
+      q = abs(denominator);
+    } else {
+      p = numerator;
+      q = denominator;
+    }
+    divider = Divider();
+  }
+
+  int Divider() {
+    if (p == q) {
+      return p;
+    } else if (p == 0 && q != 0) {
+      return q;
+    } else if (p != 0 && q == 0) {
+      return p;
+    } else {
+      int dividend_t = 0;
+      int divider_t = 0;
+      int reminder = 0;
+      if (p > q) {
+        dividend_t = p;
+        divider_t = q;
+      } else {
+        dividend_t = q;
+        divider_t = p;
+      }
+      reminder = dividend_t % divider_t;
+      while (reminder != 0) {
+        dividend_t = divider_t;
+        divider_t = reminder;
+        reminder = dividend_t % divider_t;
+      }
+      return abs(divider_t);
+    }
+  }
+
+  int Numerator() const {
+    return (p / divider);
+  }
+
+  int Denominator() const {
+    return (q / divider);
+  }
+
+ private:
+  int divider;
+  int p;
+  int q;
+};
+
+int main() {
+  {
```

Comment:
@Ypostolov Why do we need this extra scope?

Reply:
We use extra scope as access modifiers,according to the template from the task statement, we use the same name for the class instance when creating it.

Reply:
@Ypostolov According to your explanation, the main reason of this scope is using the same name for the object. Am I right?

Reply:
Yes, because from scopes depend lifetime of variables.